src/ILLink.RoslynAnalyzer/DataFlow/DynamicallyAccessedMembersBinder.cs

```diff
@@ -0,0 +1,435 @@
+						if ((@event.RemoveMethod != null) && !@event.RemoveMethod.IsStatic) continue;
+					}
+
+					if ((bindingFlags & (BindingFlags.Instance | BindingFlags.Static)) == BindingFlags.Instance) {
+						if ((@event.AddMethod != null) && @event.AddMethod.IsStatic) continue;
+						if ((@event.RemoveMethod != null) && @event.RemoveMethod.IsStatic) continue;
+					}
+
+					if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.Public) {
+						if ((@event.AddMethod == null || (@event.AddMethod.DeclaredAccessibility != Accessibility.Public))
+							&& (@event.RemoveMethod == null || (@event.RemoveMethod.DeclaredAccessibility != Accessibility.Public)))
+							continue;
+					}
+
+					if ((bindingFlags & (BindingFlags.Public | BindingFlags.NonPublic)) == BindingFlags.NonPublic) {
+						if ((@event.AddMethod != null) && @event.AddMethod.DeclaredAccessibility == Accessibility.Public) continue;
+						if ((@event.RemoveMethod != null) && @event.RemoveMethod.DeclaredAccessibility == Accessibility.Public) continue;
+					}
+
+					yield return @event;
+				}
+
+				if ((bindingFlags & BindingFlags.DeclaredOnly) == BindingFlags.DeclaredOnly)
+					yield break;
+
+				type = type.BaseType;
+				onBaseType = true;
+			}
+		}
+
+		// declaredOnly will cause this to retrieve interfaces recursively required by the type, but doesn't necessarily
+		// include interfaces required by any base types.
+		public static IEnumerable<ITypeSymbol> GetAllInterfaceImplementations (this ITypeSymbol thisType, OperationAnalysisContext context, bool declaredOnly)
+		{
+			ITypeSymbol? type = thisType;
+			while (type != null) {
+				foreach (var i in type.Interfaces) {
+					yield return i;
+
+					ITypeSymbol? interfaceType = i;
+					if (interfaceType != null) {
+						// declaredOnly here doesn't matter since interfaces don't have base types
+						foreach (var innerInterface in interfaceType.GetAllInterfaceImplementations (context, declaredOnly: true))
+							yield return innerInterface;
+					}
+				}
+
+				if (declaredOnly)
+					yield break;
+
+				type = type.BaseType;
+			}
+		}
+
+		// Issue https://github.com/dotnet/roslyn-analyzers/issues/4845
+#pragma warning disable RS1024
+		// declaredOnly will cause this to retrieve only members of the type, not of its base types. This includes interfaces recursively
+		// required by this type (but not members of these interfaces, or interfaces required only by base types).
+		public static void GetAllOnType (this ITypeSymbol type, OperationAnalysisContext context, bool declaredOnly, List<ISymbol> members) => GetAllOnType (type, context, declaredOnly, members, new HashSet<ITypeSymbol> ());
+#pragma warning restore RS1024
+
+		static void GetAllOnType (ITypeSymbol type, OperationAnalysisContext context, bool declaredOnly, List<ISymbol> members, HashSet<ITypeSymbol> types)
+		{
+			if (!types.Add (type))
+				return;
+
+			// @TODO
+			//if (type.HasNestedTypes) {
```

Comment:
Left-over TODO. I think it would be fine to do nested types as part of the loop over all members below - I don't think we rely on the order of the results.